workflow/src/main/kotlin/com/r3/corda/sdk/token/workflow/flows/RedeemToken.kt

```diff
@@ -1,85 +1,172 @@
+                    "Exactly one owned token of a particular type $ownedToken should be in the vault at any one time."
+                }
+                ownedTokenStateAndRef
+            } else {
+                val tokenSelection = TokenSelection(serviceHub)
+                val queryCriteria = ownedTokenAmountWithIssuerCriteria(amount.token, issuer)
+                val fungibleStates = tokenSelection.attemptSpend(amount, TransactionBuilder().lockId, queryCriteria) // TODO We shouldn't expose lockId in this function
+                checkSameNotary(fungibleStates)
+                fungibleStates
+            }
+            progressTracker.currentStep = SEND_STATE_REF
+            subFlow(SendStateAndRefFlow(issuerSession, exitStateAndRefs))
+
+            progressTracker.currentStep = SIGNING_TX
+            subFlow(object : SignTransactionFlow(issuerSession) {
+                // TODO Add some additional checks.
+                override fun checkTransaction(stx: SignedTransaction) = Unit
+            })
+
+            progressTracker.currentStep = FINALISING_TX
+            return subFlow(ReceiveFinalityFlow(otherSideSession = issuerSession, statesToRecord = StatesToRecord.ONLY_RELEVANT))
+        }
+    }
+
+    // Called on Issuer side.
+    @InitiatedBy(InitiateRedeem::class)
+    class IssuerResponder<T : TokenType>(val otherSession: FlowSession) : FlowLogic<SignedTransaction>() {
+        @Suspendable
+        override fun call(): SignedTransaction {
+            // Receive a redeem notification from the party. It tells us if we need to sign up for token updates or
+            // generate a confidential identity.
+            val redeemNotification = otherSession.receive<TokenRedeemNotification<T>>().unwrap { it }
+
+            // Request confidential identity, if necessary.
+            val otherIdentity = if (redeemNotification.anonymous) {
+                subFlow(RequestConfidentialIdentity.Initiator(otherSession)).party.anonymise()
+            } else otherSession.counterparty
+
+            val stateAndRefsToRedeem = subFlow(ReceiveStateAndRefFlow<AbstractToken>(otherSession))
+            check(stateAndRefsToRedeem.isNotEmpty()) {
+                "Received empty list of states to redeem."
+            }
+            checkSameIssuer(stateAndRefsToRedeem, ourIdentity)
+            checkSameNotary(stateAndRefsToRedeem)
+            checkOwner(serviceHub.identityService, stateAndRefsToRedeem)
+            val notary = stateAndRefsToRedeem.first().state.notary
+            val txBuilder = TransactionBuilder(notary = notary)
+            if (redeemNotification.amount == null) {
+                generateExitNonFungible(txBuilder, stateAndRefsToRedeem.first() as StateAndRef<NonFungibleToken<T>>)
+            } else {
+                TokenSelection(serviceHub).generateExit(txBuilder, stateAndRefsToRedeem as List<StateAndRef<FungibleToken<T>>>, redeemNotification.amount, otherIdentity)
```

Comment:
hmm maybe generate exit shouldn't be in `TokenSelection`

Reply:
It's for Cash, I guess after Stefano's token selection changes we can move it?